File: crates/harness-server/src/workflow_routes.rs
//! REST surface for declarative workflows.
//!
//! Mirrors `automation_routes.rs`: every handler 503s when
//! `state.workflows` is unset, so tests / specialised binaries that don't
//! wire the store get a clean "not configured" signal instead of a panic.
//!
//! - `GET/POST /v1/workflows` — list / create definitions.
//! - `GET/PATCH/DELETE /v1/workflows/:id` — fetch / update / delete.
//! - `POST /v1/workflows/:id/run` — dispatch a run (optionally bound to a
//!   requirement); returns `202` + the `Running` [`WorkflowRun`].
//! - `GET /v1/workflows/:id/runs` — run history for a definition.
//! - `GET /v1/workflow-runs/:run_id` — one run by id.

use axum::{
    extract::{Path, State},
    response::{IntoResponse, Response},
    routing::{get, post},
    Json, Router,
};
use harness_workflow::{WorkflowDefinition, WorkflowRun, WorkflowStep};
use serde::{Deserialize, Serialize};
use serde_json::json;

use crate::state::AppState;
use crate::workflow_runtime::execute_workflow_run;

/// Wall-clock budget for a single agent step in a manually dispatched
/// run. Matches the auto loop's `JARVIS_WORK_RUN_TIMEOUT_MS` default.
const DEFAULT_RUN_TIMEOUT_MS: u64 = 600_000;

pub(crate) fn router() -> Router<AppState> {
    Router::new()
        .route("/v1/workflows", get(list_workflows).post(create_workflow))
        .route(
            "/v1/workflows/:id",
            get(get_workflow)
                .patch(patch_workflow)
                .delete(delete_workflow),
        )
        .route("/v1/workflows/:id/run", post(run_workflow))
        .route("/v1/workflows/:id/runs", get(list_workflow_runs))
        .route("/v1/workflow-runs/:run_id", get(get_workflow_run))
}

#[derive(Debug, Serialize)]
struct WorkflowsResponse {
    items: Vec<WorkflowDefinition>,
}

#[derive(Debug, Serialize)]
struct WorkflowRunsResponse {
    items: Vec<WorkflowRun>,
}

#[derive(Debug, Deserialize)]
struct CreateWorkflowRequest {
    name: String,
    #[serde(default)]
    description: Option<String>,
    #[serde(default)]
    project_id: Option<String>,
    #[serde(default)]
    steps: Vec<WorkflowStep>,
}

#[derive(Debug, Default, Deserialize)]
struct PatchWorkflowRequest {
    #[serde(default)]
    name: Option<String>,
    /// `Some("")` clears the description; `None` leaves it as-is.
    #[serde(default)]
    description: Option<String>,
    /// `Some("")` clears the project scope; `None` leaves it as-is.
    #[serde(default)]
    project_id: Option<String>,
    #[serde(default)]
    steps: Option<Vec<WorkflowStep>>,
}

#[derive(Debug, Default, Deserialize)]
struct RunWorkflowRequest {
    #[serde(default)]
    requirement_id: Option<String>,
}

async fn list_workflows(State(state): State<AppState>) -> Response {
    let Some(store) = state.workflows.as_ref() else {
        return unavailable();
    };
    match store.list().await {
        Ok(items) => Json(WorkflowsResponse { items }).into_response(),
        Err(e) => server_error(e.to_string()),
    }
}

async fn create_workflow(
    State(state): State<AppState>,
    Json(req): Json<CreateWorkflowRequest>,
) -> Response {
    let Some(store) = state.workflows.as_ref() else {
        return unavailable();
    };
    if req.name.trim().is_empty() {
        return bad_request("name is required");
    }
    let mut def = WorkflowDefinition::new(req.name.trim());
    def.description = clean_opt(req.description);
    def.project_id = clean_opt(req.project_id);
    def.steps = normalize_steps(req.steps);
    if let Err(e) = validate_steps(&def) {
        return bad_request(e);
    }
    match store.upsert(&def).await {
        Ok(()) => (axum::http::StatusCode::CREATED, Json(def)).into_response(),
        Err(e) => server_error(e.to_string()),
    }
}

async fn get_workflow(State(state): State<AppState>, Path(id): Path<String>) -> Response {
    let Some(store) = state.workflows.as_ref() else {
        return unavailable();
    };
    match store.get(&id).await {
        Ok(Some(def)) => Json(def).into_response(),
        Ok(None) => not_found(),
        Err(e) => server_error(e.to_string()),
    }
}

async fn patch_workflow(
    State(state): State<AppState>,
    Path(id): Path<String>,
    Json(req): Json<PatchWorkflowRequest>,
) -> Response {
    let Some(store) = state.workflows.as_ref() else {
        return unavailable();
    };
    let mut def = match store.get(&id).await {
        Ok(Some(def)) => def,
        Ok(None) => return not_found(),
        Err(e) => return server_error(e.to_string()),
    };
    if let Some(name) = req.name {
        if name.trim().is_empty() {
            return bad_request("name cannot be empty");
        }
        def.name = name.trim().to_string();
    }
    if let Some(description) = req.description {
        def.description = clean_opt(Some(description));
    }
    if let Some(project_id) = req.project_id {
        def.project_id = clean_opt(Some(project_id));
    }
    if let Some(steps) = req.steps {
        def.steps = normalize_steps(steps);
    }
    if let Err(e) = validate_steps(&def) {
        return bad_request(e);
    }
    def.touch();
    match store.upsert(&def).await {
        Ok(()) => Json(def).into_response(),
        Err(e) => server_error(e.to_string()),
    }
}

async fn delete_workflow(State(state): State<AppState>, Path(id): Path<String>) -> Response {
    let Some(store) = state.workflows.as_ref() else {
        return unavailable();
    };
    match store.delete(&id).await {
        Ok(true) => axum::http::StatusCode::NO_CONTENT.into_response(),
        Ok(false) => not_found(),
        Err(e) => server_error(e.to_string()),
    }
}

async fn run_workflow(
    State(state): State<AppState>,
    Path(id): Path<String>,
    Json(req): Json<RunWorkflowRequest>,
) -> Response {
    let Some(store) = state.workflows.as_ref() else {
        return unavailable();
    };
    let def = match store.get(&id).await {
        Ok(Some(def)) => def,
        Ok(None) => return not_found(),
        Err(e) => return server_error(e.to_string()),
    };

    // Resolve an optional requirement binding for this run.
    let requirement = match clean_opt(req.requirement_id) {
        Some(req_id) => {
            let Some(req_store) = state.requirements.as_ref() else {
                return (
                    axum::http::StatusCode::SERVICE_UNAVAILABLE,
                    Json(json!({ "error": "requirement store not configured" })),
                )
                    .into_response();
            };
            match req_store.get(&req_id).await {
                Ok(Some(r)) => Some(r),
                Ok(None) => {
                    return bad_request(format!("requirement `{req_id}` not found"));
                }
                Err(e) => return server_error(e.to_string()),
            }
        }
        None => None,
    };

    // Mint + persist the Running run synchronously so we can return its
    // id, then execute in the background (à la spawn_automation_run).
    let run = WorkflowRun::new(def.id.clone(), requirement.as_ref().map(|r| r.id.clone()));
    if let Err(e) = store.upsert_run(&run).await {
        return server_error(e.to_string());
    }
    let response_run = run.clone();
    let state2 = state.clone();
    tokio::spawn(async move {
        execute_workflow_run(
            &state2,
            &def,
            requirement.as_ref(),
            run,
            None,
            DEFAULT_RUN_TIMEOUT_MS,
        )
        .await;
    });
    (axum::http::StatusCode::ACCEPTED, Json(response_run)).into_response()
}

async fn list_workflow_runs(State(state): State<AppState>, Path(id): Path<String>) -> Response {
    let Some(store) = state.workflows.as_ref() else {
        return unavailable();
    };
    match store.list_runs(Some(&id), None).await {
        Ok(items) => Json(WorkflowRunsResponse { items }).into_response(),
        Err(e) => server_error(e.to_string()),
    }
}

async fn get_workflow_run(State(state): State<AppState>, Path(run_id): Path<String>) -> Response {
    let Some(store) = state.workflows.as_ref() else {
        return unavailable();
    };
    match store.get_run(&run_id).await {
        Ok(Some(run)) => Json(run).into_response(),
        Ok(None) => not_found(),
        Err(e) => server_error(e.to_string()),
    }
}

/// Assign a fresh UUID to any step (recursively) whose id is blank, so
/// clients can POST step trees without minting ids themselves.
fn normalize_steps(steps: Vec<WorkflowStep>) -> Vec<WorkflowStep> {
    use harness_workflow::WorkflowStepKind;
    steps
        .into_iter()
        .map(|mut s| {
            if s.id.trim().is_empty() {
                s.id = uuid::Uuid::new_v4().to_string();
            }
            s.kind = match s.kind {
                WorkflowStepKind::Pipeline { steps } => WorkflowStepKind::Pipeline {
                    steps: normalize_steps(steps),
                },
                WorkflowStepKind::Phase { title, steps } => WorkflowStepKind::Phase {
                    title,
                    steps: normalize_steps(steps),
                },
                WorkflowStepKind::Parallel { steps, join } => WorkflowStepKind::Parallel {
                    steps: normalize_steps(steps),
                    join,
                },
                agent => agent,
            };
            s
        })
        .collect()
}

fn validate_steps(def: &WorkflowDefinition) -> Result<(), String> {
    if def.agent_step_count() == 0 {
        return Err("workflow must contain at least one agent step".to_string());
    }
    Ok(())
}

fn clean_opt(value: Option<String>) -> Option<String> {
    value.and_then(|s| {
        let trimmed = s.trim();
        (!trimmed.is_empty()).then(|| trimmed.to_string())
    })
}

fn unavailable() -> Response {
    (
        axum::http::StatusCode::SERVICE_UNAVAILABLE,
        Json(json!({ "error": "workflow store not configured" })),
    )
        .into_response()
}

fn bad_request(message: impl Into<String>) -> Response {
    (
        axum::http::StatusCode::BAD_REQUEST,
        Json(json!({ "error": message.into() })),
    )
        .into_response()
}

fn not_found() -> Response {
    (
        axum::http::StatusCode::NOT_FOUND,
        Json(json!({ "error": "workflow not found" })),
    )
        .into_response()
}

fn server_error(message: String) -> Response {
    (
        axum::http::StatusCode::INTERNAL_SERVER_ERROR,
        Json(json!({ "error": message })),
    )
        .into_response()
}

#[cfg(test)]
mod tests {
    use super::*;
    use harness_workflow::WorkflowStepKind;
    use std::sync::Arc;

    struct StubLlm;

    #[async_trait::async_trait]
    impl harness_core::LlmProvider for StubLlm {
        async fn complete(
            &self,
            _: harness_core::ChatRequest,
        ) -> Result<harness_core::ChatResponse, harness_core::Error> {
            Err(harness_core::Error::Provider("stub".into()))
        }
    }

    fn mk_state() -> AppState {
        let cfg = harness_core::AgentConfig::new("stub-model");
        let agent = Arc::new(harness_core::Agent::new(Arc::new(StubLlm) as _, cfg));
        AppState::new(agent).with_workflows(Arc::new(harness_store::MemoryWorkflowStore::new()))
    }

    fn agent_step(name: &str, prompt: &str) -> WorkflowStep {
        WorkflowStep::new(
            name,
            WorkflowStepKind::Agent {
                prompt: prompt.into(),
                subagent: None,
                model: None,
                output_key: None,
            },
        )
    }

    async fn read_json(resp: Response) -> serde_json::Value {
        let bytes = axum::body::to_bytes(resp.into_body(), 1024 * 1024)
            .await
            .unwrap();
        serde_json::from_slice(&bytes).unwrap()
    }

    #[tokio::test]
    async fn list_returns_503_when_store_absent() {
        let cfg = harness_core::AgentConfig::new("stub-model");
        let agent = Arc::new(harness_core::Agent::new(Arc::new(StubLlm) as _, cfg));
        let resp = list_workflows(State(AppState::new(agent))).await;
        assert_eq!(resp.status(), axum::http::StatusCode::SERVICE_UNAVAILABLE);
    }

    #[tokio::test]
    async fn create_then_list_round_trip() {
        let state = mk_state();
        let create = create_workflow(
            State(state.clone()),
            Json(CreateWorkflowRequest {
                name: "ship".into(),
                description: Some("research then build".into()),
                project_id: None,
                steps: vec![agent_step("research", "look around")],
            }),
        )
        .await;
        assert_eq!(create.status(), axum::http::StatusCode::CREATED);
        let created = read_json(create).await;
        assert_eq!(created["name"], "ship");
        // The blank-id step got a server-minted id.
        assert!(!created["steps"][0]["id"].as_str().unwrap().is_empty());

        let list = list_workflows(State(state)).await;
        let body = read_json(list).await;
        assert_eq!(body["items"].as_array().unwrap().len(), 1);
    }

    #[tokio::test]
    async fn create_rejects_workflow_with_no_agent_steps() {
        let state = mk_state();
        let resp = create_workflow(
            State(state),
            Json(CreateWorkflowRequest {
                name: "empty".into(),
                description: None,
                project_id: None,
                steps: vec![],
            }),
        )
        .await;
        assert_eq!(resp.status(), axum::http::StatusCode::BAD_REQUEST);
    }

    #[tokio::test]
    async fn run_dispatches_and_persists_running_run() {
        let state = mk_state();
        // Create a definition first.
        let create = create_workflow(
            State(state.clone()),
            Json(CreateWorkflowRequest {
                name: "ship".into(),
                description: None,
                project_id: None,
                steps: vec![agent_step("research", "look around")],
            }),
        )
        .await;
        let def = read_json(create).await;
        let wf_id = def["id"].as_str().unwrap().to_string();

        let run = run_workflow(
            State(state.clone()),
            Path(wf_id.clone()),
            Json(RunWorkflowRequest {
                requirement_id: None,
            }),
        )
        .await;
        assert_eq!(run.status(), axum::http::StatusCode::ACCEPTED);
        let run_body = read_json(run).await;
        let run_id = run_body["id"].as_str().unwrap().to_string();
        assert_eq!(run_body["workflow_id"], wf_id);

        // The Running run was persisted synchronously before the spawn.
        let got = get_workflow_run(State(state), Path(run_id)).await;
        assert_eq!(got.status(), axum::http::StatusCode::OK);
    }
}

File: packages/server/src/workflow-routes.ts
// REST surface for declarative workflows. Ported from
// harness-server/src/workflow_routes.rs (the CRUD + run-ledger subset).
//
// Every handler 503s when `state.workflows` is unset, so tests / specialised
// binaries that don't wire the store get a clean "not configured" signal
// instead of a panic — mirrors the conversations-routes `requireStore` pattern.
//
//   - GET/POST  /v1/workflows                 — list / create definitions.
//   - GET/PUT/DELETE /v1/workflows/:id         — fetch / update / delete.
//     (PATCH is also accepted as an alias for the update verb, since the Rust
//      source uses PATCH; both map to the same handler.)
//   - POST /v1/workflows/:id/run               — dispatch a run (optionally
//     bound to a requirement); returns 202 + the new WorkflowRun.
//   - GET  /v1/workflows/:id/runs              — run history for a definition.
//   - GET  /v1/workflow-runs/:run_id           — one run by id.
//   - POST /v1/workflow-runs/:run_id/cancel    — cancel an in-flight run.
//
// DEFERRAL (P-later): the Rust route reserves a process-wide concurrency slot
// (429 on exhaustion) and spawns `execute_workflow_run` in the background. The
// actual execution engine (workflow_runtime) + the concurrency gate / reaper
// land in a later phase. Until a runtime driver exists, `POST /:id/run` just
// persists a *Pending* run and returns 202 — nothing executes it yet, and
// there is no concurrency cap to enforce. `cancel` likewise has no live task to
// abort; it only finalises the persisted row.
import type { FastifyInstance, FastifyReply } from "fastify";
import { errorText } from "@jarvis/core";
import {
  agentStepCount,
  finishWorkflowRun,
  newWorkflowDefinition,
  newWorkflowStep,
  touchDefinition,
  WorkflowError,
  workflowRunStatusIsTerminal,
  type WorkflowDefinition,
  type WorkflowRun,
  type WorkflowStep,
  type WorkflowStepKind,
  type WorkflowStore,
} from "@jarvis/workflow";
import { randomUUID } from "node:crypto";
import type { AppState } from "./state.ts";

// ---------- request bodies ----------

interface CreateWorkflowRequest {
  name?: string;
  description?: string;
  project_id?: string;
  steps?: WorkflowStep[];
}

interface PatchWorkflowRequest {
  name?: string;
  /** `""` clears the description; absent leaves it as-is. */
  description?: string;
  /** `""` clears the project scope; absent leaves it as-is. */
  project_id?: string;
  steps?: WorkflowStep[];
}

interface RunWorkflowRequest {
  requirement_id?: string;
}

// ---------- store guard ----------

/** Return the workflow store, or send a 503 and return undefined. */
function requireStore(state: AppState, reply: FastifyReply): WorkflowStore | undefined {
  if (!state.workflows) {
    reply.code(503).send({ error: "workflow store not configured" });
    return undefined;
  }
  return state.workflows;
}

// ---------- helpers ----------

/** Trim a string, returning undefined when it's empty/whitespace-only. */
function cleanOpt(value: string | undefined): string | undefined {
  if (value === undefined) return undefined;
  const trimmed = value.trim();
  return trimmed.length > 0 ? trimmed : undefined;
}

/**
 * Validate the wire shape of a step tree, returning a human-readable error
 * string (or `null` when well-formed). `normalizeSteps` reads `s.kind.type`
 * and recurses into container `steps`, so a malformed payload (non-array, a
 * non-object step, or a step missing its `kind`) would throw a raw TypeError
 * → 500. Checking the shape first lets the routes answer 400 (see issue #415).
 */
function stepsShapeError(steps: unknown): string | null {
  if (!Array.isArray(steps)) return "steps must be an array";
  for (const s of steps) {
    if (s === null || typeof s !== "object") return "each step must be an object";
    const kind = (s as { kind?: unknown }).kind;
    if (kind === null || typeof kind !== "object") return "each step requires a `kind` object";
    const type = (kind as { type?: unknown }).type;
    if (typeof type !== "string") return "each step's `kind.type` must be a string";
    if (type === "pipeline" || type === "phase" || type === "parallel") {
      const nested = stepsShapeError((kind as { steps?: unknown }).steps);
      if (nested) return nested;
    }
  }
  return null;
}

/**
 * Assign a fresh UUID to any step (recursively) whose id is blank, so clients
 * can POST step trees without minting ids themselves. Mirrors `normalize_steps`.
 */
function normalizeSteps(steps: WorkflowStep[]): WorkflowStep[] {
  return steps.map((s) => {
    const id = s.id != null && s.id.trim().length > 0 ? s.id : randomUUID();
    let kind: WorkflowStepKind = s.kind;
    if (kind.type === "pipeline") {
      kind = { type: "pipeline", steps: normalizeSteps(kind.steps) };
    } else if (kind.type === "phase") {
      kind = { type: "phase", title: kind.title, steps: normalizeSteps(kind.steps) };
    } else if (kind.type === "parallel") {
      kind = { type: "parallel", steps: normalizeSteps(kind.steps), join: kind.join };
    }
    return { id, name: s.name, kind };
  });
}

/**
 * Validate a definition: it must carry at least one agent step, and every
 * agent step's prompt must be non-empty. Throws {@link WorkflowError} (whose
 * `.message` matches the Rust `#[error(...)]` text). Mirrors `validate_steps`.
 */
function validateSteps(def: WorkflowDefinition): void {
  if (agentStepCount(def) === 0) {
    throw WorkflowError.empty();
  }
  const walk = (steps: WorkflowStep[]): void => {
    for (const step of steps) {
      const kind = step.kind;
      if (kind.type === "agent") {
        if (kind.prompt.trim().length === 0) {
          throw WorkflowError.emptyPrompt(step.name);
        }
      } else {
        walk(kind.steps);
      }
    }
  };
  walk(def.steps);
}

// ---------- routes ----------

export function registerWorkflowRoutes(app: FastifyInstance, state: AppState): void {
  // list
  app.get("/v1/workflows", async (_req, reply) => {
    const store = requireStore(state, reply);
    if (!store) return reply;
    try {
      const items = await store.list();
      return reply.send({ items });
    } catch (e) {
      return reply.code(500).send({ error: errorText(e) });
    }
  });

  // create
  app.post("/v1/workflows", async (req, reply) => {
    const store = requireStore(state, reply);
    if (!store) return reply;
    const body = (req.body ?? {}) as CreateWorkflowRequest;
    const name = typeof body.name === "string" ? body.name : "";
    if (name.trim().length === 0) {
      return reply.code(400).send({ error: "name is required" });
    }
    // Guard optional fields before `cleanOpt`/`normalizeSteps` dereference them:
    // a mistyped `description`/`project_id`/`steps` would otherwise throw a raw
    // TypeError → 500 instead of a clean 400 (see issue #415).
    if (body.description !== undefined && typeof body.description !== "string") {
      return reply.code(400).send({ error: "description must be a string" });
    }
    if (body.project_id !== undefined && typeof body.project_id !== "string") {
      return reply.code(400).send({ error: "project_id must be a string" });
    }
    if (body.steps !== undefined) {
      const stepsErr = stepsShapeError(body.steps);
      if (stepsErr) return reply.code(400).send({ error: stepsErr });
    }
    const def = newWorkflowDefinition(name.trim());
    def.description = cleanOpt(body.description);
    def.project_id = cleanOpt(body.project_id);
    def.steps = normalizeSteps(body.steps ?? []);
    try {
      validateSteps(def);
    } catch (e) {
      if (e instanceof WorkflowError) return reply.code(400).send({ error: e.message });
      throw e;
    }
    try {
      await store.upsert(def);
    } catch (e) {
      return reply.code(500).send({ error: errorText(e) });
    }
    return reply.code(201).send(def);
  });

  // get one
  app.get("/v1/workflows/:id", async (req, reply) => {
    const store = requireStore(state, reply);
    if (!store) return reply;
    const id = (req.params as { id: string }).id;
    try {
      const def = await store.get(id);
      if (!def) return reply.code(404).send({ error: "workflow not found" });
      return reply.send(def);
    } catch (e) {
      return reply.code(500).send({ error: errorText(e) });
    }
  });

  // update — Rust uses PATCH; the task asks for PUT. Register both pointing at
  // the same handler so the wire contract holds either way.
  const updateHandler = async (req: { params: unknown; body: unknown }, reply: FastifyReply) => {
    const store = requireStore(state, reply);
    if (!store) return reply;
    const id = (req.params as { id: string }).id;
    let def: WorkflowDefinition | undefined;
    try {
      def = await store.get(id);
    } catch (e) {
      return reply.code(500).send({ error: errorText(e) });
    }
    if (!def) return reply.code(404).send({ error: "workflow not found" });
    const body = (req.body ?? {}) as PatchWorkflowRequest;
    // Guard each optional field's type before dereferencing (issue #415):
    // `name`/`description`/`project_id` feed `.trim()`/`cleanOpt`, and `steps`
    // feeds `normalizeSteps` — a mistyped value would throw a raw TypeError → 500.
    if (body.name !== undefined) {
      if (typeof body.name !== "string" || body.name.trim().length === 0) {
        return reply.code(400).send({ error: "name cannot be empty" });
      }
      def.name = body.name.trim();
    }
    if (body.description !== undefined) {
      if (typeof body.description !== "string") {
        return reply.code(400).send({ error: "description must be a string" });
      }
      def.description = cleanOpt(body.description);
    }
    if (body.project_id !== undefined) {
      if (typeof body.project_id !== "string") {
        return reply.code(400).send({ error: "project_id must be a string" });
      }
      def.project_id = cleanOpt(body.project_id);
    }
    if (body.steps !== undefined) {
      const stepsErr = stepsShapeError(body.steps);
      if (stepsErr) return reply.code(400).send({ error: stepsErr });
      def.steps = normalizeSteps(body.steps);
    }
    try {
      validateSteps(def);
    } catch (e) {
      if (e instanceof WorkflowError) return reply.code(400).send({ error: e.message });
      throw e;
    }
    touchDefinition(def);
    try {
      await store.upsert(def);
    } catch (e) {
      return reply.code(500).send({ error: errorText(e) });
    }
    return reply.send(def);
  };
  app.put("/v1/workflows/:id", updateHandler);
  app.patch("/v1/workflows/:id", updateHandler);

  // delete
  app.delete("/v1/workflows/:id", async (req, reply) => {
    const store = requireStore(state, reply);
    if (!store) return reply;
    const id = (req.params as { id: string }).id;
    try {
      const deleted = await store.delete(id);
      if (!deleted) return reply.code(404).send({ error: "workflow not found" });
      return reply.code(204).send();
    } catch (e) {
      return reply.code(500).send({ error: errorText(e) });
    }
  });

  // dispatch a run. DEFERRAL: the execution engine (workflow_runtime) + the
  // concurrency gate land later. For now we persist a *Pending* run and 202;
  // nothing executes it yet.
  app.post("/v1/workflows/:id/run", async (req, reply) => {
    const store = requireStore(state, reply);
    if (!store) return reply;
    const id = (req.params as { id: string }).id;
    let def: WorkflowDefinition | undefined;
    try {
      def = await store.get(id);
    } catch (e) {
      return reply.code(500).send({ error: errorText(e) });
    }
    if (!def) return reply.code(404).send({ error: "workflow not found" });

    // Resolve an optional requirement binding for this run.
    const body = (req.body ?? {}) as RunWorkflowRequest;
    const reqId = cleanOpt(body.requirement_id);
    let requirementId: string | undefined;
    if (reqId !== undefined) {
      if (!state.requirements) {
        return reply.code(503).send({ error: "requirement store not configured" });
      }
      let requirement;
      try {
        requirement = await state.requirements.get(reqId);
      } catch (e) {
        return reply.code(500).send({ error: errorText(e) });
      }
      if (!requirement) {
        return reply.code(400).send({ error: `requirement \`${reqId}\` not found` });
      }
      requirementId = requirement.id;
    }

    // Mint a Pending run. (Rust mints `running` + spawns the executor; without
    // a runtime driver we leave it Pending so a later phase can pick it up.)
    const run: WorkflowRun = {
      id: randomUUID(),
      workflow_id: def.id,
      status: "pending",
      step_results: [],
      started_at: new Date().toISOString(),
    };
    if (requirementId !== undefined) run.requirement_id = requirementId;
    try {
      await store.upsertRun(run);
    } catch (e) {
      return reply.code(500).send({ error: errorText(e) });
    }
    return reply.code(202).send(run);
  });

  // run history for a definition
  app.get("/v1/workflows/:id/runs", async (req, reply) => {
    const store = requireStore(state, reply);
    if (!store) return reply;
    const id = (req.params as { id: string }).id;
    try {
      const items = await store.listRuns(id, undefined);
      return reply.send({ items });
    } catch (e) {
      return reply.code(500).send({ error: errorText(e) });
    }
  });

  // one run by id
  app.get("/v1/workflow-runs/:run_id", async (req, reply) => {
    const store = requireStore(state, reply);
    if (!store) return reply;
    const runId = (req.params as { run_id: string }).run_id;
    try {
      const run = await store.getRun(runId);
      if (!run) return reply.code(404).send({ error: "workflow run not found" });
      return reply.send(run);
    } catch (e) {
      return reply.code(500).send({ error: errorText(e) });
    }
  });

  // cancel an in-flight run. DEFERRAL: no live task to abort yet — we only
  // finalise the persisted row. Idempotent on already-terminal runs.
  app.post("/v1/workflow-runs/:run_id/cancel", async (req, reply) => {
    const store = requireStore(state, reply);
    if (!store) return reply;
    const runId = (req.params as { run_id: string }).run_id;
    let run: WorkflowRun | undefined;
    try {
      run = await store.getRun(runId);
    } catch (e) {
      return reply.code(500).send({ error: errorText(e) });
    }
    if (!run) return reply.code(404).send({ error: "workflow run not found" });
    if (workflowRunStatusIsTerminal(run.status)) {
      // Already finished — nothing to abort. Idempotent success.
      return reply.send(run);
    }
    if (run.error === undefined) {
      run.error = "run cancelled by operator";
    }
    finishWorkflowRun(run, "cancelled");
    try {
      await store.upsertRun(run);
    } catch (e) {
      return reply.code(500).send({ error: errorText(e) });
    }
    return reply.send(run);
  });
}

// Re-export so the test file (and Assemble step) can reference the helper used
// to mint step trees without re-importing from @jarvis/workflow.
export { newWorkflowStep };
